src/plasmapy/simulation/particle_tracker/particle_tracker.py

```diff
@@ -413,6 +436,67 @@ def load_particles(
         self.x = x.to(u.m).value
         self.v = v.to(u.m / u.s).value
 
+    def add_Bethe_stopping(self, I: u.Quantity):  # noqa: E741
+        r"""
+        Enable particle stopping described by non-relativistic Bethe formula.
+
+        The electron density associated with each material must be provided in
+        the grid during instantiation.
+
+        Parameters
+        ----------
+        I : `~astropy.units.Quantity`, shape (ngrids)
+            An array of the mean excitation energy values for each of the provided
+            grids.
+
+        Raises
+        ------
+        `ValueError`:
+            The provided mean excitation energy array does not match with the number
+            of grids provided at class instantiation.
+        """
+
+        if len(I) != len(self.grids):
+            raise ValueError(
+                "Please provide an array of length ngrids for the mean excitation energies."
+            )
+
+        self._do_stopping = True
+
+        # Require that each grid has a defined electron density,
+        # if a grid does not define n_e, raise an exception
+        for grid in self.grids:
+            grid.require_quantities("n_e", replace_with_zeros=False)
+
+    def add_stopping(self, materials: list[str]):
+        r"""
+        Enable particle stopping using experimental stopping powers.
+
+        Interpolation of stopping powers is conducted using data from the NIST
+        PSTAR. This information is combined with the mass density quantity
+        provided in the grids to calculate the energy loss over the distance
+        travelled during a timestep.
+        """
+
+        if len(materials) != len(self.grids):
+            raise ValueError(
+                "Please provide an array of length ngrids for the materials."
+            )
+
+        # Require that each grid has a defined mass density,
+        # if a grid does not define rho, raise an exception
+        # TODO: should this be replaced with only raising an exception when none
```

Comment:
Remember this function replaces any missing values with zero arrays, so it depends how you handle it later. Probably the fastest way is to not require it, but then only do the stopping calculation if it is present? 

I would raise a warning here if NO grid has rho...